rasa/utils/tensorflow/model_data_utils.py

```diff
@@ -1,226 +1,370 @@
+
+
+TAG_ID_ORIGIN = "tag_id_origin"
+
+
+def convert_training_examples(
+    training_examples: List[Message],
+    attributes: List[Text],
+    entity_tag_specs: Optional[List["EntityTagSpec"]] = None,
+    featurizers: Optional[List[Text]] = None,
+    bilou_tagging: bool = False,
+) -> List[Dict[Text, List["Features"]]]:
+    """Converts training data into a list of attribute to features.
+
+    Possible attributes are, for example, INTENT, RESPONSE, TEXT, ACTION_TEXT,
+    ACTION_NAME or ENTITIES.
+
+    Args:
+        training_examples: the list of training examples
+        attributes: the attributes to consider
+        entity_tag_specs: the entity specs
+        featurizers: the featurizers to consider
+        bilou_tagging: indicates whether BILOU tagging should be used or not
+
+    Returns:
+        A list of attribute to features.
+    """
+    output = []
+
+    for example in training_examples:
+        attribute_to_features = {}
+        for attribute in attributes:
+            if attribute == ENTITIES:
+                attribute_to_features[attribute] = []
+                # in case of entities add the tag_ids
+                for tag_spec in entity_tag_specs:
+                    attribute_to_features[attribute].append(
+                        _get_tag_ids(example, tag_spec, bilou_tagging)
+                    )
+            elif attribute in example.data:
+                attribute_to_features[attribute] = example.get_all_features(
+                    attribute, featurizers
+                )
+        output.append(attribute_to_features)
+
+    return output
+
+
+def _get_tag_ids(
+    example: Message, tag_spec: "EntityTagSpec", bilou_tagging: bool
+) -> "Features":
+    """Creates a feature array containing the tag ids of the given example."""
+    from rasa.nlu.test import determine_token_labels
+    from rasa.nlu.utils.bilou_utils import bilou_tags_to_ids
+    from rasa.shared.nlu.training_data.features import Features
+
+    if bilou_tagging:
+        _tags = bilou_tags_to_ids(example, tag_spec.tags_to_ids, tag_spec.tag_name)
+    else:
+        _tags = []
+        for token in example.get(TOKENS_NAMES[TEXT]):
+            _tag = determine_token_labels(
+                token, example.get(ENTITIES), attribute_key=tag_spec.tag_name
+            )
+            _tags.append(tag_spec.tags_to_ids[_tag])
 
-MASK = "mask"
+    # transpose to have seq_len x 1
+    return Features(
+        np.array([_tags]).T, FEATURE_TYPE_SEQUENCE, tag_spec.tag_name, TAG_ID_ORIGIN
+    )
 
 
-def surface_attributes(
-    tracker_state_features: List[List[Dict[Text, List["Features"]]]]
+def _surface_attributes(
+    features: List[List[Dict[Text, List["Features"]]]],
+    featurizers: Optional[List[Text]] = None,
 ) -> Dict[Text, List[List[List["Features"]]]]:
     """Restructure the input.
 
+    "features_for_examples" can, for example, be a dictionary of attributes (INTENT,

```

Comment:
Same comment regarding renaming `features_for_examples` as below. Also, the explanation of `features_for_examples` already exists in `convert_to_data_format`, so maybe we can skip that and describe how are we restructuring/what's the schema of the output.

Reply:
@tabergma could you please take this over